### PR TITLE
NumPy: adding `size`, `shape_ptr`, `stride_ptr`

### PR DESCRIPTION
**Pull request by tdegeus**
Fixes https://github.com/wjakob/nanobind/discussions/158

I was not sure how you deal with tests. I could not find tests for e.g. `stride` to I tried to be non-intrusive. But let me know if you want otherwise.

### CODE REVIEW

File: include/nanobind/ndarray.h [hunk 1]
```diff
@@ -293,9 +293,12 @@ template <typename... Args> class ndarray {
     }
 
     dlpack::dtype dtype() const { return m_dltensor.dtype; }
+    size_t size() const { return Info::size; }
```
Comment (wjakob) [inline include/nanobind/ndarray.h:296]: Isn't `size()` doing the same as `ndim()` here? I would have expected `size()` to return the total number of tensor coefficients.

File: include/nanobind/ndarray.h [hunk 2]
```diff
@@ -293,9 +293,12 @@ template <typename... Args> class ndarray {
...
     size_t ndim() const { return m_dltensor.ndim; }
     size_t shape(size_t i) const { return m_dltensor.shape[i]; }
     int64_t stride(size_t i) const { return m_dltensor.strides[i]; }
+    auto* shape_ptr() const { return &m_dltensor.shape[0]; }
```
Comment (wjakob) [inline include/nanobind/ndarray.h:300]: Let's not use `auto` in a public interface especially when the type is easily knowable. You can return the array directly and don't need `&array[0]`.

**Review by wjakob — CHANGES_REQUESTED** [review]
A quick review. Two further requests:

- Please at least reference those functions somewhere from a C++ test case.
- They will still need to be mentioned in the API documentation `docs/api_extra.rst`

File: include/nanobind/ndarray.h (hunk 2 above)
Comment (tdegeus) [reply include/nanobind/ndarray.h]: Done. I'm just a bit confused about the type of `shape_ptr`. From the other return types I would have expected it to be `size_t*`. However, my compiler is convinced about it being `int64_t*`

**Comment by wjakob** [comment]
The changed code looks good, but please note the request above regarding documentation + test coverage.

**Comment by wjakob** [comment]
Thanks!

**Pull request merged**